synapse/storage/_base.py

```diff
@@ -301,17 +303,32 @@ def runInteraction(self, desc, func, *args, **kwargs):
 
         def inner_func(conn, *args, **kwargs):
             with LoggingContext("runInteraction") as context:
-                sql_scheduling_timer.inc_by(time.time() * 1000 - start_time)
+                sched_delay_ms = time.time() * 1000 - start_time
+                sql_scheduling_timer.inc_by(sched_delay_ms)
 
                 if self.database_engine.is_connection_closed(conn):
                     logger.debug("Reconnecting closed database connection")
                     conn.reconnect()
 
                 current_context.copy_to(context)
-                return self._new_transaction(
-                    conn, desc, after_callbacks, final_callbacks, current_context,
-                    func, *args, **kwargs
-                )
+                txn_start_time_ms = time.time() * 1000
+                try:
+                    return self._new_transaction(
+                        conn, desc, after_callbacks, final_callbacks, current_context,
+                        func, *args, **kwargs
+                    )
+                finally:
+                    txn_end_time_ms = time.time() * 1000
+                    txn_duration = txn_end_time_ms - txn_start_time_ms
+
+                    current_context.add_database_transaction(
+                        txn_duration, sched_delay_ms,
+                    )
+                    self._current_txn_total_time += txn_duration
+                    self._txn_perf_counters.update(
+                        desc, txn_start_time_ms, txn_end_time_ms,
+                    )
+                    sql_txn_timer.inc_by(txn_duration, desc)
```

Comment:
Won't this mean that any transactions created by calling runWithConnection directly won't be measured?

Reply:
well, yes, but they aren't currently measured as part of `db_txn_duration`.

Looks like the only thing that uses `runWithConnection` other than the background updates is the `event_fetch` code, which is not specific to any http requests. I'd like for time spent waiting for events to be fetched to be tracked somehow, but tracking transaction duration here wouldn't help.

Reply:
They may not be picked up by `db_txn_duration`, but they are picked up by `sql_txn_timer`. Losing metrics for fetching event txn duration doesn't sound ideal.